File: packages/dcf_primitives/ios/Classes/Components/DCFSpinnerComponent.swift
/*
 * Copyright (c) Dotcorr Studio. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */


import UIKit
import dcflight

class DCFSpinnerComponent: NSObject, DCFComponent {
    private static let sharedInstance = DCFSpinnerComponent()
    
    required override init() {
        super.init()
    }
    
    func createView(props: [String: Any]) -> UIView {
        
        let spinner: UIActivityIndicatorView
        
        let style = props["style"] as? String ?? "medium"
        
        if #available(iOS 13.0, *) {
            switch style.lowercased() {
            case "large":
                spinner = UIActivityIndicatorView(style: .large)
            case "small":
                spinner = UIActivityIndicatorView(style: .medium)
            default: // medium
                spinner = UIActivityIndicatorView(style: .medium)
            }
        } else {
            switch style.lowercased() {
            case "large":
                spinner = UIActivityIndicatorView(style: .whiteLarge)
            case "small":
                spinner = UIActivityIndicatorView(style: .white)
            default: // medium
                spinner = UIActivityIndicatorView(style: .gray)
            }
        }
        
        if let spinnerColor = ColorUtilities.getColor(
            explicitColor: "spinnerColor",
            semanticColor: "primaryColor",
            from: props
        ) {
            spinner.color = spinnerColor
        }
        
        updateView(spinner, withProps: props)
        spinner.applyStyles(props: props)
        
        return spinner
    }
    
    func updateView(_ view: UIView, withProps props: [String: Any]) -> Bool {
        guard let spinner = view as? UIActivityIndicatorView else { return false }
        
        if let animating = props["animating"] as? Bool {
            if animating {
                spinner.startAnimating()
            } else {
                spinner.stopAnimating()
            }
        } else {
            spinner.startAnimating()
        }
        
        if let spinnerColor = ColorUtilities.getColor(
            explicitColor: "spinnerColor",
            semanticColor: "primaryColor",
            from: props
        ) {
            spinner.color = spinnerColor
        }
        
        if let hidesWhenStopped = props["hidesWhenStopped"] as? Bool {
            spinner.hidesWhenStopped = hidesWhenStopped
        } else {
            spinner.hidesWhenStopped = true
        }
        
        spinner.applyStyles(props: props)
        return true
    }
    
    func applyLayout(_ view: UIView, layout: YGNodeLayout) {
        view.frame = CGRect(x: layout.left, y: layout.top, width: layout.width, height: layout.height)
    }

    func viewRegisteredWithShadowTree(_ view: UIView, shadowView: DCFShadowView, nodeId: String) {
        objc_setAssociatedObject(view,
                               UnsafeRawPointer(bitPattern: "nodeId".hashValue)!,
                               nodeId,
                               .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
    }

    static func handleTunnelMethod(_ method: String, params: [String: Any]) -> Any? {
        return nil
    }
}

File: packages/dcf_primitives/android/src/main/kotlin/com/dotcorr/dcf_primitives/components/DCFSpinnerComponent.kt
/*
 * Copyright (c) Dotcorr Studio. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.dotcorr.dcf_primitives.components

import android.content.Context
import android.content.res.ColorStateList
import android.graphics.Color
import android.graphics.PointF
import android.view.View
import android.widget.ProgressBar
import com.dotcorr.dcflight.components.DCFComponent
import com.dotcorr.dcflight.components.DCFTags
import com.dotcorr.dcflight.extensions.applyStyles
import com.dotcorr.dcflight.utils.ColorUtilities

class DCFSpinnerComponent : DCFComponent() {

    override fun createView(context: Context, props: Map<String, Any?>): View {
        val progressBar = ProgressBar(context)

        progressBar.isIndeterminate = true

        ColorUtilities.getColor("spinnerColor", "primaryColor", props)?.let { colorInt ->
            progressBar.indeterminateTintList = ColorStateList.valueOf(colorInt)
        }

        progressBar.setTag(DCFTags.COMPONENT_TYPE_KEY, "Spinner")

        updateView(progressBar, props)
        return progressBar
    }

    override fun updateView(view: View, props: Map<String, Any?>): Boolean {
        val progressBar = view as? ProgressBar ?: return false
        
        // CRITICAL: Merge new props with existing stored props to preserve all properties
        val existingProps = getStoredProps(view)
        val mergedProps = mergeProps(existingProps, props)
        storeProps(view, mergedProps)
        
        val nonNullProps = mergedProps.filterValues { it != null }.mapValues { it.value!! }

        val isAnimating = mergedProps["animating"] as? Boolean ?: true
        val hidesWhenStopped = mergedProps["hidesWhenStopped"] as? Boolean ?: true

        if (hidesWhenStopped) {
            progressBar.visibility = if (isAnimating) View.VISIBLE else View.GONE
        } else {
            progressBar.visibility = View.VISIBLE
        }

        ColorUtilities.getColor("spinnerColor", "primaryColor", nonNullProps)?.let { colorInt ->
            progressBar.indeterminateTintList = ColorStateList.valueOf(colorInt)
        }

        mergedProps["size"]?.let { size ->
            when (size) {
                "small", "medium" -> {
                    progressBar.scaleX = 0.7f
                    progressBar.scaleY = 0.7f
                }
                "large" -> {
                    progressBar.scaleX = 1.2f
                    progressBar.scaleY = 1.2f
                }
                else -> {
                    progressBar.scaleX = 1.0f
                    progressBar.scaleY = 1.0f
                }
            }
        }

        view.applyStyles(nonNullProps)
        return true
    }


    override fun viewRegisteredWithShadowTree(view: View, shadowNode: com.dotcorr.dcflight.layout.DCFShadowNode, nodeId: String) {
    }

    override fun handleTunnelMethod(method: String, arguments: Map<String, Any?>): Any? {
        return null
    }
}

